README.md
Chore
=====

Chore is a system health monitoring tool for everyone.

There are more sophisticated tools, tools with better disaster
recovery, tools that monitor CPU/Swap/Disk-space, to be sure.  But
these usually require switching gears from your main codebase, an
advanced learning curve, and often find themselves used only by a
dedicated systems person or a single resident expert on your team.

Chore provides a simple API that anyone can start using in 5 minutes
or less.  This allows any developer to add their own monitoring to
tasks with little to no effort.

Chore provides a simple ruby interface to indicate:

*   When a task is started.  (Is it actually running?)

*   Optionally when it finished and/or errored.  (Is it completing?)

*   Optionally parameters such as how frequently the job should run
    and/or how long it should take.  (Is it behaving as expected?)

Chore does not:

*   Perform automated recovery of failed tasks.

*   Kill or restart zombie processes.

*   Provide server health info such as high CPU usage, low disk space,
    excessive swapping, etc.

To start a server:

    chore-server

This will start a server.

To view status from the command line:

    chore-status [hostname] [port]

Which will produce output like:

    crazy_background_task - started 2012-06-11 17:11:16 -0400 (Job should run every 1 minute, but has a grace period of 40 minutes)
    random_resque_job - started 2012-06-11 17:11:16 -0400 (Should run every 20 minutes)
    custom_script - started 2012-06-11 17:11:16 -0400 (Should run every 1 hour)
    logrotate - started 2012-06-11 17:11:16 -0400 (Job should run every 1 second, but hasn't run since 2012-06-11 17:11:16 -0400)
    exceptional - failed 2012-06-11 17:11:16 -0400 (Another freaking nil error)
    exceptionally_anonymous - failed 2012-06-11 17:11:16 -0400 (FAILED!!!)
    finish_anytime - finished 2012-06-11 17:11:16 -0400 (no particular deadline)
    slow - finished 2012-06-11 17:11:16 -0400 (Finished)
    quick - finished 2012-06-11 17:11:16 -0400 (Finished, but 4 seconds late!!!)
    good task - finished 2012-06-11 17:11:21 -0400 (no particular deadline)
    bad task - failed 2012-06-11 17:11:21 -0400 (RuntimeError - AAAAAAAAAAAAAAAAAAAAA)

(Todo, how do we show the colors in github)

To view status from a web-server:

    http://localhost:43210

To record a chore in ruby:

```ruby

# one time config, not needed if server is on localhost
Chore.set_server('hostname',Chore::Constants::DEFAULT_LISTEN_PORT)

Chore.monitor(:task_name) do
  # real work
end
```

This will record the start and finish of the wrapped block.

Even simpler, you can just record the start of a task:

```ruby
Chore.start(:task_name)
#do stuff
```

This still requires you to examine timestamps to see if things are
running.  If you want a better visual clue that things are broken, you
can throw in some options.  (All .start options can also be passed
into .monitor)

```ruby
# complain if the task hasn't run in more than an hour.
Chore.start(:task_name, :do_every => 3600)
    
# complain if the task hasn't run in more than an hour, and
# complain more loudly if it hasn't run in two.
Chore.start(:task_name, :do_every => 3600, :grace_period => 3600)
    
# complain if the task hasn't finished in an hour.
Chore.start(:task_name, :finish_in => 3600)
# make sure to call Chore.finish(:task_name) explicitly later, 
# or use Chore.monitor to do so automatically.    

# Record error info for known exception
begin
  Chore.start(:goofy_task)
  raise Errno::ETIMEDOUT # network problems
rescue Errno::ETIMEDOUT => ex
  Chore.fail(:goofy_task, :error => "Crappy router probably needs a reboot."
end

# Add status notes, only the last update shows.
Chore.start(:long_task)
# ...
Chore.status(:long_task, "Downloaded the interwebz")
# ...
Chore.status(:long_task, "Ran bayesian classifier")
# ...

# Remove a task from the store on completion, but not if it fails
Chore.monitor("update widget #{widget.id}", :pop => true) do
  # ...
end

# Expire an entry in X seconds, whether it's started, finished or
# failed, so that it no longer shows up in output.  Useful when
# attaching an id number or pid to the task name.
Chore.monitor("Cron task with pid #{process_id}", :expire_in => 1.day) {}

```

There is also a command-line tool so you can use the server without
having to fire up irb.

```
johnmudhead:chore grant$ chore-status
test - failed 2012-06-15 17:33:00 -0400 (AAAAAAAAA)
johnmudhead:chore grant$ chore-client --chore test --action pop
johnmudhead:chore grant$ chore-status

johnmudhead:chore grant$ 
```

Verification
------------

This gem is signed with rubygems-openpgp.  You can verify its
integrity by running:

    gem install chore --trust

Much more information on signing is available at the [rubygems-openpgp
Certificate Authority](https://www.rubygems-openpgp-ca.org).


Signing key:

    pub   2048R/E3B5806F 2010-01-11 [expires: 2014-01-03]
          Key fingerprint = A530 C31C D762 0D26 E2BA  C384 B6F6 FFD0 E3B5 806F
    uid                  Grant T. Olson (Personal email) <kgo@grant-olson.net>
    uid                  Grant T Olson <grant@webkite.com>
    uid                  Grant T. Olson (pikimal) <grant@pikimal.com>
    sub   2048R/6A8F7CF6 2010-01-11 [expires: 2014-01-03]
    sub   2048R/A18A54D6 2010-03-01 [expires: 2014-01-03]
    sub   2048R/D53982CE 2010-08-31 [expires: 2014-01-03]


HISTORY
**** 0.2.4 - March 28, 2013 ****

* Don't use past tense in status so we don't see 'updateed'.

* Show first line of traceback when .monitor catches a failure.

**** 0.2.3 - September 20, 2012 ****

* Always display text indicating that a job hasn't been running on
  schedule, even if it has failed for other reasons.

* Config files so travis-ci will run.

**** 0.2.2 - July 3, 2012 ****

* Chore.send now truncates unusually large messages that are too big
  for UDP, and the server will fail gracefully if receiving such a message
  fron and out of date client.

**** 0.2.1 - June 22, 2012 ****

* CLI interface now has a --refresh option, which gives top-style
  updates.

* Web interface now auto-refreshes every 60 seconds.

* CLI interface now supports all client options.

**** 0.2.0 - June 15, 2012 ****

Initial public release.

chore-server.rb
require 'eventmachine'
require 'evma_httpserver'
require 'json'
require './chore_time'

module ChoreStore
  @@store = {}

  def self.get
    @@store
  end

  def self.colorize str, color
    color_code = case color
                 when :red then 31
                 when :green then 32
                 when :yellow then 33
                 else raise "BAD COLOR #{str} #{color}"
                 end
    "\033[#{color_code}m#{str}\033[0m"
  end

  def self.iterate_statuses
    ret = []
    ChoreStore.get.each_pair do |key, val|
      puts val.inspect
      status = val['status'].to_sym
      run_time = val['start_time']
      run_time = 0 if !run_time

      current_time = Time.now.to_i
      do_every = val['do_every']
      grace_period = val['grace_period']

      notes = []
      state = :red

      if status == :fail
        state = :red
        if val['error']
          notes << val['error']
        else
          notes << "FAILED!!!"
        end
        
      elsif status == :finish
        finish_time = val['finish_time']
        finish_in = val['finish_in']

        if finish_in.nil?
          state = :green
          notes << "no particular deadline"
        elsif (run_time + finish_in) >= finish_time
          state = :green
          notes << "Finished"
        else
          state = :red
          notes << "Finished, but #{finish_time - (run_time + finish_in)} seconds late!!!"
        end
      elsif status == :start
        if do_every
          if run_time + do_every >= current_time
            state = :green
            notes << "Should run every #{ChoreTime.elapsed_human_time(do_every)}"
          elsif grace_period && run_time + do_every + grace_period > current_time
            state = :yellow
            notes << "Job should run every #{ChoreTime.elapsed_human_time(do_every)}, but has a grace period of #{ChoreTime.elapsed_human_time(grace_period)}"
          else
            state = :red
            notes << "Job should run every #{ChoreTime.elapsed_human_time(do_every)}, but hasn't run since #{Time.at(run_time)}"
          end
        else
          state = :green
          notes << "No regular schedule."
        end
      end
      
      info = {:job => key, :state => state, :status => status, :start_time => run_time, :notes => notes}
      yield info

    end
  end

  def self.text_statuses

    status_lines = []
    iterate_statuses do |status|
      status_line = "#{status[:job]} - #{status[:status]}ed #{Time.at(status[:start_time])}"
      status_line += " (#{status[:notes].join(', ')})" if !status[:notes].empty?
      status_lines << colorize(status_line, status[:state])
    end
    
    status_lines.join("\n") + "\n"
  end
  
  #
  # Process data with a spawned process in the background 
  @@data_collector = EM.spawn do |chore_info|
    start_or_finish = chore_info[0]
    chore = chore_info[1]
    opts = chore_info[2]
    opts['status'] = start_or_finish
    
    if ChoreStore.get[chore].nil?
      ChoreStore.get[chore] = {}
    end

    ChoreStore.get[chore] = ChoreStore.get[chore].merge(opts)
    
  end

  def self.collect chore_info
    @@data_collector.notify chore_info
  end

end


module ChoreCollect
  def receive_data(data)
    chore_info = JSON.parse(data)
    ChoreStore.collect chore_info
  end
end

module ChoreDisplay
  def receive_data(data)
    send_data(ChoreStore.text_statuses)
    close_connection_after_writing
  end
end

class ChoreWeb < EventMachine::Connection
  include EventMachine::HttpServer

  def process_http_request
    resp = EventMachine::DelegatedHttpResponse.new(self)

    html = <<-html
<html>
<head>
<style type="text/css">
body {font-family:monospace;background-color:#CCCCCC;}
.red {color:red;}
.yellow {color:yellow;}
.green {color:green;}
table, th, td { border: 1px solid black;}
</style>
</head>
<body>
<h1>Chores</h1>
<table>
<tr><th>Job</th><th>Status</th><th>Time</th><th>Notes</th></tr>
html

    ChoreStore.iterate_statuses do |status|
      row = "<tr class='#{status[:state]}'><td>#{status[:job]}</td><td>#{status[:status]}ed</td><td>#{Time.at(status[:start_time])}</td>"
      if !status[:notes].empty?
        row += "<td>(#{status[:notes].join(', ')})</td>"
      else
        row += "<td>&nbsp;</td>"
      end
      
      row += "</tr>\n"
      html << row 
    end

    html << "</body></html>"

    resp.status = 200
    resp.content = html
    resp.send_response
  end
  
end


EventMachine::run do
  EventMachine::open_datagram_socket('0.0.0.0', 7779, ChoreCollect)
  EventMachine::start_server('0.0.0.0', 8888, ChoreDisplay)
  EventMachine::start_server('0.0.0.0', 8899, ChoreWeb)
end
